src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -1,280 +1,886 @@
+        }
+
+        position->open_part             = position->open_part->next;
+        bytes_remaining                 = F_PART_DATA_SIZE - position->part_offset;
+    }
+    //bytes_to_write < bytes_remaining
+    memcpy(position->open_part->data, buffer, bytes_to_write);
+    position->part_offset               += bytes_to_write;
+    position->file_offset               += bytes_to_write;
+    if (o->size < position->file_offset) {
+        o->size                         = position->file_offset;
+    }
+    return nbytes;
 }
 
-uint32 file_insert(struct fsobj *o, struct FS *filesys)
+struct fsobj *file_find(char *path)
 {
-    assert(filesys && filesys->root && filesys->root->filedes);
-    if (filesys->root == o) return OS_FS_SUCCESS;
+    assert(path);
+    //paths should always begin with '/' dir names do not
+    if (path[0] != '/') return NULL;
+    path++;
+
+    // token is the current directory in path
+    char path_temp[OS_MAX_PATH_LEN * 2];
+    strcpy(path_temp, path);
+    const char delim[2] = "/";
+    char *token = strtok(path_temp, delim);
+
+    if(!openfs || !openfs->root) return NULL;
+    struct fsobj *cur = openfs->root;
+    assert(cur && cur->name);
+    if (strcmp(token, cur->name) != 0) return NULL;
+
+    while (1) {
+        // iterate through linked list of children until ancestor is found
+        while (strcmp(token, cur->name) != 0) {
+            if (!cur->next) return NULL;
+            cur = cur->next;
+        }
+        token = strtok(NULL, delim);
+        if (token == NULL) return cur;
+        if (!cur->child) return NULL;
+        cur = cur->child;
+    }
+    PANIC("Unreachable Statement");
+    return NULL;
+}
 
-    insert(o, filesys->root);
+int32 file_create(char *path, permission_t permission)
+{
+    assert(path);
+
+    struct fsobj *o;
+    if (newfile_get(&o) != OS_FS_SUCCESS) return OS_FS_ERROR;
+
+    o->name = path_to_name(path);
+    o->type = FSOBJ_FILE;
+    o->size = 0;
+    o->permission = permission;
+
+    struct f_part *part;
+    newpart_get(&part);
+    assert(part);
+
+    o->memtype = DYNAMIC;
+    o->file_part = part;
+    part->file = o;
+    part->data = (char *) part + sizeof(struct f_part);
+    if (file_insert(o, path) != OS_FS_SUCCESS){
+       o->ino = 0;
+       return OS_FS_ERR_PATH_INVALID;
+    }
+    return file_open(path, permission);
+}
 
+int32 file_remove(char *path)
+{
+    struct fsobj *file = file_find(path);
+    if (!file) return OS_FS_ERR_PATH_INVALID;
+    file_rm(file);
     return OS_FS_SUCCESS;
 }
 
-//could be written iteratively to use less stackframe if that becomes an issue
-// Root is
-uint32 insert(struct fsobj *o, struct fsobj *root)
+int32 file_rename(char *old_filename, char *new_filename)
 {
-    assert(root && o);
-    assert(strlen(root->name) < strlen(o->name));
+    struct fsobj *file = file_find(old_filename);
+    if (!file) return OS_FS_ERR_PATH_INVALID;
+    file->name = path_to_name(new_filename);
```

Comment:
the existance of a file of the same name as new_filename is checked for in the wrapper layer